README.md
[![Build Status](http://ci.digitalbazaar.com/buildStatus/icon?job=bedrock-ledger-http)](http://ci.digitalbazaar.com/job/bedrock-ledger-http)

# Bedrock Ledger HTTP API

A [bedrock][] module to control the creation and management of decentralized
ledgers via an HTTP API.

## Requirements

- npm v3+

## Quick Examples

```
npm install bedrock-ledger-http
```

```js
var actor = 'admin';

var ledgerEndpoint = 'https://example.org/ledgers/';

var ledgerConfigEvent = {
  '@context': 'https://w3id.org/flex/v1',
  id: 'did:c02915fc-672d-4568-8e6e-b12a0b35cbb3/events/1',
  type: 'LedgerConfigurationEvent',
  ledgerConfig: {
    id: 'did:c02915fc-672d-4568-8e6e-b12a0b35cbb3',
    type: 'LedgerConfiguration',
    name: 'test-ledger',
    description: 'A test ledger',
    storageMechanism: 'SequentialList',
    consensusAlgorithm: {
      type: 'ProofOfSignature2016',
      approvedSigner: [ 'https://example.org/keys/authorized-1' ],
      minimumSignaturesRequired: 1
    },
  },
  previousEvent: {
    hash: 'urn:sha256:0000000000000000000000000000000000000000000000000000000000000000';
  }
};

jsigs.sign(ledgerConfigurationEvent, {
    algorithm: 'LinkedDataSignature2015',
    privateKeyPem: myPrivateKey,
    creator: 'https://example.org/keys/authorized-1'
  }, function(err, signedConfigEvent) {
    if(err) {
      return console.log('Signature failed:', err);
    }
    request.post({
      url: ledgerEndpoint,
      body: signedConfigEvent,
      json: true
    }, function(err, res, body) {
      if(err) {
        console.log('Ledger creation failed!');
      } else if(res.statusCode === 201) {
        console.log('Ledger created!')
      }
    });
  });
```

## Configuration

For documentation on configuration, see [config.js](./lib/config.js).

## API

### POST /ledgers

Create a new ledger

Schema: [services.ledger.postConfig](./schemas/services.ledger.js)

 * Response codes:
    * 201: Ledger creation was successful. HTTP Location header contains URL of newly created ledger.
    * 400: Ledger creation failed due to malformed request.
    * 403: Ledger creation failed due to invalid digital signature.
    * 409: Ledger creation failed due to duplicate information.

### POST /ledgers/{ledger}

Append a new event to a ledger.

Schema: [services.ledger.postLedgerEvent](./schemas/services.ledger.js)

  * Response Codes:
    * 201: Ledger was appended successfully. HTTP Location header contains URL of newly appended event.
    * 400: Ledger append failed due to malformed request.
    * 403: Ledger append failed due to invalid digital signature
    * 409: Ledger append failed due to duplicate information.

### GET /ledgers

Get metadata for all ledgers known to the system.

  * Response Codes:
    * 200: [Successful response](./docs/examples/get.ledgers.jsonld)

### GET /ledgers/{ledger}

Get metadata for specific ledger known to the system.

  * Response Codes:
    * 200: [Successful response](./docs/examples/get.ledger.jsonld)
    * 404: Ledger not found.

### GET /ledgers/{ledger}/state

Get state machine information for a particular object.

  * Response Codes:
    * 200: [Successful response](./docs/examples/get.ledger.state.jsonld)
    * 400: Retrieval failed due to malformed query.
    * 404: Query generated zero results; object not found.

### GET /ledgers/{ledger}/{ledgerId}/events/{eventNumber}

Get metadata for specific ledger event.

  * Response Codes:
    * 200: [Successful response](./docs/examples/get.ledger.event.jsonld)
    * 404: Ledger event was not found.

[bedrock]: https://github.com/digitalbazaar/bedrock


CHANGELOG.md
# bedrock-ledger-http ChangeLog

## 0.4.1 - 2017-05-11

### Changed
- Update test suite.
- Apply ES6 Syntax.

## 0.4.0 - 2017-05-05

### Added
- Implement ledger agent concept.

## 0.3.5 - 2017-05-04

### Fixed
- Fix documentation build bug.

## 0.3.4 - 2017-05-03

### Fixed
- Upgrade test architecture, fix minor bugs.

## 0.2.0 - 2016-12-03

- See git history for changes previous to this release.


test/test.js
/*
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const bedrock = require('bedrock');

require('bedrock-express');
require('bedrock-identity-http');
require('bedrock-key-http');
require('bedrock-ledger-http');

require('bedrock-test');

bedrock.start();


tests/test.js
/*
 * Copyright (c) 2017 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

var bedrock = require('bedrock');

require('bedrock-express');
require('bedrock-requirejs');
require('bedrock-validation');
require('bedrock-views');
require('bedrock-identity');
require('bedrock-identity-http');
require('bedrock-key');
require('bedrock-key-http');
require('bedrock-ledger-http');

require('bedrock-test');
bedrock.start();


tests/helpers.js
/*
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
 /* jshint node: true */
'use strict';

var _ = require('lodash');
var async = require('async');
var brKey = require('bedrock-key');
var brIdentity = require('bedrock-identity');
var config = require('bedrock').config;
var uuid = require('node-uuid').v4;
var database = require('bedrock-mongodb');

var api = {};
module.exports = api;

api.createIdentity = function(userName) {
  var userId = config.server.baseUri + config['identity-http'].basePath + '/' +
    userName;
  var newIdentity = {
    id: userId,
    type: 'Identity',
    sysSlug: userName,
    label: userName,
    email: userName + '@dhs2016ledger.dev',
    sysPassword: 'password',
    sysPublic: ['label', 'url', 'description'],
    sysResourceRole: [],
    url: config.server.baseUri,
    description: userName
  };
  return newIdentity;
};

api.createKeyPair = function(options) {
  var userName = options.userName;
  var publicKey = options.publicKey;
  var privateKey = options.privateKey;
  var keyId = config.server.baseUri + config.key.basePath + '/' +
    userName + '-key-1';
  var ownerId = null;
  if(userName === 'userUnknown') {
    ownerId = '';
  } else {
    ownerId = options.userId;
  }
  var newKeyPair = {
    publicKey: {
      '@context': 'https://w3id.org/identity/v1',
      id: keyId,
      type: 'CryptographicKey',
      owner: ownerId,
      label: userName + ' signing key',
      publicKeyPem: publicKey
    },
    privateKey: {
      type: 'CryptographicKey',
      owner: ownerId,
      label: userName + ' signing key',
      publicKey: keyId,
      privateKeyPem: privateKey
    }
  };
  return newKeyPair;
};

api.prepareDatabase = function(mockData, callback) {
  async.series([
    function(callback) {
      api.removeCollections(callback);
    },
    function(callback) {
      insertTestData(mockData, callback);
    }
  ], callback);
};

api.removeCollections = function(callback) {
  var collectionNames =
    ['ledger', 'dhs2016ledger_ledger'];
  database.openCollections(collectionNames, function(err) {
    async.each(collectionNames, function(collectionName, callback) {
      database.collections[collectionName].remove({}, callback);
    }, function(err) {
      callback(err);
    });
  });
};

api.removeCollection = function(collection, callback) {
  var collectionNames = [collection];
  database.openCollections(collectionNames, function(err) {
    async.each(collectionNames, function(collectionName, callback) {
      database.collections[collectionName].remove({}, callback);
    }, function(err) {
      callback(err);
    });
  });
};

// Insert identities and public keys used for testing into database
function insertTestData(mockData, callback) {
  // add to view variables
  config.views.vars['dhs2016ledger'] = {
    identity: {},
    authorizedSigners: []
  };

  async.forEachOf(mockData.identities, function(identity, key, callback) {
    async.parallel([
      function(callback) {
        brIdentity.insert(null, identity.identity, callback);
        var viewsIdentity = identity.identity;
        viewsIdentity.publicKey = identity.keys.publicKey;
        viewsIdentity.privateKey = identity.keys.privateKey;
        config.views.vars['dhs2016ledger'].identity[viewsIdentity.sysSlug] =
          viewsIdentity;
        if(viewsIdentity.sysSlug != 'isis') {
          config.views.vars['dhs2016ledger'].authorizedSigners.push(
            viewsIdentity.publicKey.id);
        }
      },
      function(callback) {
        brKey.addPublicKey(null, identity.keys.publicKey, callback);
      }
    ], callback);
  }, function(err) {
    if(err) {
      if(!database.isDuplicateError(err)) {
        // duplicate error means test data is already loaded
        return callback(err);
      }
    }
    callback();
  }, callback);
}


lib/index.js
/*
 * Ledger HTTP API module
 *
 * Copyright (c) 2016-2017 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
const brPassport = require('bedrock-passport');
const brRest = require('bedrock-rest');
const cors = require('cors');
const docs = require('bedrock-docs');
const ledger = require('bedrock-ledger');

const BedrockError = bedrock.util.BedrockError;
const ensureAuthenticated = brPassport.ensureAuthenticated;
const validate = require('bedrock-validation').validate;

const logger = bedrock.loggers.get('app');

require('./config');

// add routes
bedrock.events.on('bedrock-express.configure.routes', addRoutes);

function addRoutes(app) {
  const basePath = bedrock.config.ledger.basePath;

  app.options(basePath, cors());

  // create a new ledger agent
  app.post(basePath,
    validate('services.ledger.postConfig'), (req, res, next) => {
      const ledgerConfigEvent = req.body;
      // FIXME: add permissions in real product
      ledger.createLedger(null, ledgerConfigEvent, {}, (err, ledgerUrl) => {
        if(err) {
          return next(err);
        }
        // return the saved config
        res.location(ledgerUrl);
        res.status(201).end();
      });
    });
  docs.annotate.post(basePath, {
    description: 'Create a new ledger agent',
    schema: 'services.ledger.postConfig',
    securedBy: ['null'],
    responses: {
      201: 'Ledger agent creation was successful. HTTP Location header ' +
        'contains URL of newly created ledger agent.',
      400: 'Ledger agent creation failed due to malformed request.',
      403: 'Ledger agent creation failed due to invalid digital signature.',
      409: 'Ledger agent creation failed due to duplicate information.'
    }
  });

  // request that a ledger agent writes an event to the ledger
  app.post(basePath + '/:agent/events',
    validate('services.ledger.postLedgerEvent'), (req, res, next) => {
      const ledgerName = req.params.agent;
      const ledgerEvent = req.body;
      // FIXME: add permissions in real product
      ledger.writeLedgerEvent(
        null, ledgerName, ledgerEvent, {}, (err, eventUrl) => {
          if(err) {
            return next(err);
          }
          // return the saved config
          res.location(eventUrl);
          res.status(201).end();
        });
    });
  docs.annotate.post(basePath + '/:agent/events', {
    description: 'Request that a Ledger Agent append a new event to a ledger',
    schema: 'services.ledger.postLedgerEvent',
    securedBy: ['null'],
    responses: {
      201: 'Ledger event was accepted for writing. HTTP Location header ' +
        'contains URL of accepted event.',
      400: 'Ledger append request failed due to malformed request.',
      403: 'Ledger append request failed due to invalid digital signature',
      409: 'Ledger append request failed due to duplicate information.'
    }
  });

  // get information on all ledger agents known to this system
  app.get(basePath,
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: (req, res, callback) => {
        const actor = req.user ? req.user.identity : undefined;

        // TODO: Implement owner-based querying
        if(req.query.owner) {
          return callback(new BedrockError('Query by owner not implemented'));
        }

        ledger.getAllLedgerMetadata(actor, {}, (err, ledgers) => {
          if(err) {
            return callback(err);
          }
          callback(err, ledgers);
        });
      }
    })
  );
  docs.annotate.get(basePath, {
    description: 'Get information on all ledger agents known to this system.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledgers.jsonld'
        }
      }
    }
  });

  // get status information for a specific ledger agent on the system
  app.get(basePath + '/:agent',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: (req, res, callback) => {
        const actor = req.user ? req.user.identity : undefined;
        const ledgerName = req.params.agent;
        ledger.getLedgerMetadata(
          actor, ledgerName, {}, (err, ledgerMetadata) => {
            if(err) {
              return callback(err);
            }
            callback(err, ledgerMetadata);
          });
      }
    })
  );
  docs.annotate.get(basePath + '/:agent', {
    description:
      'Get status information for specific ledger agent known to this system.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.jsonld'
        }
      },
      404: 'Ledger not found.'
    }
  });

  // query the current state machine associated with a ledger
  app.get(basePath + '/:agent/query',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: (req, res, callback) => {
        if(!req.query.id) {
          return callback(new BedrockError(
            'Query missing \'id\' query parameter.',
            'BadRequest', {
              httpStatusCode: 400,
              public: true
            }
          ));
        }
        const actor = req.user ? req.user.identity : undefined;
        const ledgerName = req.params.agent;
        const objectId = req.query.id;
        ledger.getStateMachineObject(
          actor, ledgerName, objectId, {}, (err, item) => {
            if(err) {
              return callback(err);
            }
            callback(err, item);
          });
      }
    })
  );
  docs.annotate.get(basePath + '/:agent/query', {
    description: 'Query the current state machine associated with a ledger.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.state.jsonld'
        }
      },
      400: 'Request failed due to malformed query.',
      404: 'Query generated zero results; object not found.'
    }
  });

  // get a specific block in the system
  app.get(basePath + '/:agent/blocks',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: (req, res, callback) => {
        const actor = req.user ? req.user.identity : undefined;
        const ledgerName = req.params.agent;
        const eventId = req.query.id;
        ledger.getLedgerEvent(
          actor, ledgerName, eventId, {}, (err, ledgerEventMetadata) => {
            if(err) {
              return callback(err);
            }
            callback(err, ledgerEventMetadata);
          });
      }
    })
  );
  docs.annotate.get(basePath + '/:agent/blocks', {
    description: 'Get information about a specific block in the ledger.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.event.jsonld'
        }
      },
      404: 'Ledger block was not found.'
    }
  });
}


lib/test.config.js
/*
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */

var config = require('bedrock').config;
var path = require('path');

config.mocha.tests.push(path.join(__dirname, '..', 'tests'));

// MongoDB
config.mongodb.name = 'dhs2016ledger_test';
config.mongodb.host = 'localhost';
config.mongodb.port = 27017;
config.mongodb.local.collection = 'dhs2016ledger_test';
config.mongodb.username = 'admin';
config.mongodb.password = 'password';
config.mongodb.adminPrompt = true;
config.mongodb.dropCollections.onInit = true;
config.mongodb.dropCollections.collections = [];

// server info
config.server.port = 39443;
config.server.httpPort = 39080;
config.server.bindAddr = ['dhs2016ledger.dev'];
config.server.domain = 'dhs2016ledger.dev';
config.server.host = 'dhs2016ledger.dev:39080';
config.server.baseUri = 'http://' + config.server.host;

//only run the local package tests
config.mocha.tests = ['tests/01-http-api.js'];

schemas/services.ledger.js
/*
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
var constants = require('bedrock').config.constants;
var schemas = require('bedrock-validation').schemas;

var postConfig = {
  title: 'Ledger Configuration Event',
  description: 'An object that describes the configuration for a ledger.',
  type: 'object',
  properties: {
    '@context': schemas.jsonldContext(constants.FLEX_LEDGER_CONTEXT_V1_URL),
    id: schemas.identifier({required: false}),
    type: schemas.jsonldType('LedgerConfigurationEvent'),
    ledgerConfig: {
      type: 'object',
      properties: {
        id: schemas.identifier({required: false}),
        type: schemas.jsonldType('LedgerConfiguration'),
        name: {
          title: 'Ledger Name',
          description: 'A human-readable short name for a ledger.',
          required: true,
          type: 'string'
        },
        description: {
          title: 'Ledger Description',
          description: 'A human-readable description of the purpose of a ledger.',
          required: true,
          type: 'string'
        },
        storageMechanism: schemas.jsonldType('SequentialList'),
        consensusAlgorithm: {
          type: 'object',
          properties: {
            type: schemas.jsonldType('ProofOfSignature2016'),
            approvedSigner: {
              type: 'array',
              items: {
                type: schemas.url()
              }
            },
            minimumSignaturesRequired: {
              type: 'number',
              minimum: 1,
              maximum: 7
            }
          }
        },
      },
    },
    previousEvent: {
      type: 'object',
      properties: {
        hash: {
          title: 'Block hash',
          description: 'The hash value the previous event in the blockchain.',
          required: true,
          type: schemas.url()
        },
      },
    },
    signature: schemas.linkedDataSignature()
  }
};

var postLedgerEvent = {
  type: 'object',
  properties: {
    '@context': schemas.jsonldContext(constants.FLEX_LEDGER_CONTEXT_V1_URL),
    type: schemas.jsonldType('LedgerStorageEvent'),
    replacesObject: {
      type: 'array',
      items: {
        type: 'object'
      }
    },
    previousEvent: {
      type: 'object',
      properties: {
        id: schemas.identifier({required: true}),
        hash: {
          title: 'Block hash',
          description: 'The hash value the previous event in the blockchain.',
          required: true,
          type: schemas.url()
        },
      },
    },
    signature: schemas.linkedDataSignature()
  }
};

module.exports.postConfig = function() {
  return postConfig;
};

module.exports.postLedgerEvent = function() {
  return postLedgerEvent;
};


lib/ledger.http.js
/*
 * Ledger HTTP API module
 *
 * Copyright (c) 2016 Digital Bazaar, Inc. All rights reserved.
 */
var _ = require('lodash');
var async = require('async');
var bedrock = require('bedrock');
var brPassport = require('bedrock-passport');
var brRest = require('bedrock-rest');
var config = require('bedrock').config;
var database = require('bedrock-mongodb');
var cors = require('cors');
var docs = require('bedrock-docs');
var ledger = require('./ledger.storage');
var util = require('util');

var BedrockError = bedrock.util.BedrockError;
var ensureAuthenticated = brPassport.ensureAuthenticated;
var validate = require('bedrock-validation').validate;

var logger = bedrock.loggers.get('app');

// add routes
bedrock.events.on('bedrock-express.configure.routes', addRoutes);

function addRoutes(app) {
  var basePath = bedrock.config.ledger.basePath;

  app.options(basePath, cors());

  // create new ledger
  app.post(basePath,
    validate('services.ledger.postConfig'),
    function(req, res, next) {
      var ledgerConfigEvent = req.body;
      // FIXME: add permissions in real product
      ledger.createLedger(
        null, ledgerConfigEvent, {}, function(err, ledgerUrl) {
        if(err) {
          return next(err);
        }
        // return the saved config
        res.set('Location', ledgerUrl);
        res.status(201).end();
      });
  });
  docs.annotate.post(basePath, {
    description: 'Create a new ledger',
    schema: 'services.ledger.postConfig',
    securedBy: ['null'],
    responses: {
      201: 'Ledger creation was successful. HTTP Location header contains ' +
        'URL of newly created ledger.',
      400: 'Ledger creation failed due to malformed request.',
      403: 'Ledger creation failed due to invalid digital signature.',
      409: 'Ledger creation failed due to duplicate information.'
    }
  });

  // write information to a ledger
  app.post(basePath + '/:ledger',
    validate('services.ledger.postLedgerEvent'),
    function(req, res, next) {
      var ledgerName = req.params.ledger;
      var ledgerEvent = req.body;
      // FIXME: add permissions in real product
      ledger.writeLedgerEvent(
        null, ledgerName, ledgerEvent, {}, function(err, eventUrl) {
        if(err) {
          return next(err);
        }
        // return the saved config
        res.set('Location', eventUrl);
        res.status(201).end();
      });
  });
  docs.annotate.post(basePath + '/:ledger', {
    description: 'Append a new event to a ledger',
    schema: 'services.ledger.postLedgerEvent',
    securedBy: ['null'],
    responses: {
      201: 'Ledger was appended successfully. HTTP Location header contains ' +
        'URL of newly appended event.',
      400: 'Ledger append failed due to malformed request.',
      403: 'Ledger append failed due to invalid digital signature',
      409: 'Ledger append failed due to duplicate information.'
    }
  });

  // get metadata for all ledgers in the system
  app.get(basePath,
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var actor = req.user ? req.user.identity : undefined;
        ledger.getAllLedgerMetadata(actor, {}, function(err, ledgers) {
          if(err) {
            return callback(err);
          }
          callback(err, ledgers);
        });
      }
    })
  );
  docs.annotate.get(basePath, {
    description: 'Get metadata for all ledgers known to the system.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledgers.jsonld'
        }
      }
    }
  });

  // get metadata for specific ledger on the system
  app.get(basePath + '/:ledger',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var actor = req.user ? req.user.identity : undefined;
        var ledgerName = req.params.ledger;
        ledger.getLedgerMetadata(
          actor, ledgerName, {}, function(err, ledgerMetadata) {
          if(err) {
            return callback(err);
          }
          callback(err, ledgerMetadata);
        });
      }
    })
  );
  docs.annotate.get(basePath + '/:ledger', {
    description: 'Get metadata for specific ledger known to the system.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.jsonld'
        }
      },
      404: 'Ledger not found.'
    }
  });

  // get state machine information for a particular object
  app.get(basePath + '/:ledger/state',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        if(!req.query.id) {
          return callback(new BedrockError(
            'Query missing \'id\' query parameter.',
            'BadRequest', {
              httpStatusCode: 400,
              public: true
            }
          ));
        }
        var actor = req.user ? req.user.identity : undefined;
        var ledgerName = req.params.ledger;
        var objectId = req.query.id;
        ledger.getStateMachineObject(
          actor, ledgerName, objectId, {}, function(err, item) {
          if(err) {
            return callback(err);
          }
          callback(err, item);
        });
      }
    })
  );
  docs.annotate.get(basePath + '/:ledger/state', {
    description: 'Get state machine information for a particular object.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.state.jsonld'
        }
      },
      400: 'Retrieval failed due to malformed query.',
      404: 'Query generated zero results; object not found.'
    }
  });

  // get metadata for specific ledger on the system
  app.get(basePath + '/:ledger/:ledgerId/events/:eventNumber',
    cors(),
    brRest.when.prefers.ld,
    brRest.linkedDataHandler({
      get: function(req, res, callback) {
        var actor = req.user ? req.user.identity : undefined;
        var ledgerName = req.params.ledger;
        var eventId = req.params.ledgerId + '/events/' + req.params.eventNumber;
        ledger.getLedgerEvent(
          actor, ledgerName, eventId, {}, function(err, ledgerEventMetadata) {
          if(err) {
            return callback(err);
          }
          callback(err, ledgerEventMetadata);
        });
      }
    })
  );
  docs.annotate.get(basePath + '/:ledger/:ledgerId/events/:eventNumber', {
    description: 'Get metadata for specific ledger event.',
    securedBy: ['null'],
    responses: {
      200: {
        'application/ld+json': {
          example: 'examples/get.ledger.event.jsonld'
        }
      },
      404: 'Ledger event was not found.'
    }
  });
}
